packages/bvaughn-architecture-demo/src/suspense/TokenizerCache.ts

```diff
@@ -0,0 +1,69 @@
+import { javascriptLanguage } from "@codemirror/lang-javascript";
+import { ensureSyntaxTree } from "@codemirror/language";
+import { EditorState } from "@codemirror/state";
+import { Tree } from "@lezer/common";
+import { highlightTree, classHighlighter } from "@lezer/highlight";
+
+import { createGenericCache } from "./createGenericCache";
+
+function codeToState(code: string): EditorState {
+  return EditorState.create({
+    doc: code,
+    extensions: [javascriptLanguage.extension],
+  });
+}
+
+function stateToTree(state: EditorState): Tree | null {
+  return ensureSyntaxTree(state, state.doc.length, 1e9);
+}
+
+function identity(any: any) {
+  return any;
+}
+
+// TODO
+// Suspense can be async; we could move this to a Worker if it's slow.
+async function tokenizer(code: string): Promise<Tree | null> {
+  const state = codeToState(code);
+  const tree = stateToTree(state);
+
+  return tree;
+}
+
+// TODO
+// Suspense can be async; we could move this to a Worker if it's slow.
+async function highlighter(code: string, tree: Tree): Promise<string[]> {
```

Comment:
we can also be smart and operate on a range of lines as well. This would probably work pretty well if we're already returning an array of lines as well.

This is only needed when we're trying to be lazy and not highlight the entire document.

Reply:
Heh, my main immediate question about the approach is here is that we'd need to start implementing a virtualized list ourselves :)

Reply:
Yeah for sure. I know exactly how I'm going to window this 😄